### PR TITLE
[RC-1453] fx-ify the remote config service component

### PR DESCRIPTION
**Pull request by moth-a**
This is a re-opening #22347 which was previously merged. The PR was reverted by #22431). The previous PR introduced two unintended behaviors which have now been fixed:
* One where the agent would fail to start entirely if the RC service failed to initialize. This has been fixed in [this new commit ](https://github.com/DataDog/datadog-agent/pull/22726/commits/318982df37ad4358760bd75418115a21b4a3f41c)
* Another where the FX lifecycle context that was being passed to the RC service' was being cancelled 5 minutes after application startup, causing premature stoppage of the RC service. We were previously relying on context cancellations as a mechanism to control the service's lifecycle, but that has now been fixed in [this new commit](https://github.com/DataDog/datadog-agent/pull/22726/commits/325765fb82c237e6f153fcf0196a5ad7f5344a78)

Lastly, I have since [introduced some e2e tests](https://github.com/DataDog/datadog-agent/pull/22677) covering the startup behavior of the agent and the RC service component to ensure this changes are now behaving as expected ([tests are passing for this PR](https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/432014924)).

### What does this PR do?

This PR refactors a bit of remote config service initialization code that was polluting `cmd/agent/common`. There are no functional changes. To accomplish this, the rc service is now being provided as an uber fx component to the agent commands that need it.

### Motivation

First, there is an overarching initiative to migrate to using uber.fx to supply various components throughout the agent.

Second, the remote config service initialization logic was pulled out to `cmd/agent/common` (by me) during a prior refactor which is an antipattern, and it did not belong there

Third, as part of the HAMR initiative, we will need to conditionally instantiate/inject a second RC service that communicates with the R2 failover datacenter. Since the RC service depends on an RCTelemetryReporter which can only be initialized/configured once, we really need a singleton telemetry reporter anyways, and uber.fx is the right way to do that regardless. By componentizing everything like this, it sets us up to provide the same telemetry reporter singleton to both services in an upcoming/unrelated PR.

### Describe how to test/QA your changes

- start the cluster agent with RC disabled via `remote_configuration.enabled: false` in the agent config. Ensure the cluster agent starts fine and that remote config is indeed disabled
- start the cluster agent with RC enabled (including a valid API Key etc). Ensure the cluster agent starts fine, and that remote config is indeed enabled
- start the core agent with RC disabled via `remote_configuration.enabled: false` in the agent config. Ensure the cluster agent starts fine and that remote config is indeed disabled
- start the core agent with RC enabled (including a valid API Key etc). Ensure the cluster agent starts fine, and that remote config is indeed enabled
- start the core agent with the following config (be sure NOT to explicitly set the `remote_configuration.no_tls_validation` config field). This should cause an error during remote config service initialization. Ensure the agent starts fine without the remote config service and logs a corresponding error:
```
skip_ssl_validation: true
remote_configuration:
  enabled: true
```
```
2024-01-30 13:20:32 EST | CORE | ERROR | (comp/remote-config/rcservice/rcserviceimpl/rcservice.go:61 in newRemoteConfigServiceOptional) | remote config service not initialized or started: unable to create remote-config service: remote Configuration does not allow skipping TLS validation by default (currently skipped because `skip_ssl_validation` is set to true). While it is not advised, the `remote_configuration.no_tls_validation` config option can be set to `true` to disable this protection
```

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/config/remote/service/service.go [hunk 1]
```diff
@@ -305,22 +308,23 @@ func newRCBackendOrgUUIDProvider(http api.API) uptane.OrgUUIDProvider {
...
+func (s *Service) Start() {
 	go func() {
 		s.pollOrgStatus()
 		for {
 			select {
 			case <-s.clock.After(orgStatusPollInterval):
 				s.pollOrgStatus()
-			case <-ctx.Done():
+			case <-s.stopOrgPoller:
+				log.Infof("Remote Config service context cancelled. Stopping org status poller")
```
Comment (BaptisteFoy) [inline pkg/config/remote/service/service.go:319]: **🥜 Nitpick:** ‏Log's inaccurate

**Review by BaptisteFoy — APPROVED** [review]

**Review by jennchenn — APPROVED** [review]

**Review by pducolin — APPROVED** [review]

**Comment by pducolin** [comment]
Ok for the e2e test, you can move `test/new-e2e/tests/remote-config` code ownership to remote config team

File: cmd/cluster-agent/subcommands/start/command.go [hunk 2]
```diff
@@ -198,17 +202,15 @@ func start(log log.Component, config config.Component, taggerComp tagger.Compone
 
 	// Initialize remote configuration
 	var rcClient *rcclient.Client
-	var rcService *rcservice.Service
-	if pkgconfig.IsRemoteConfigEnabled(pkgconfig.Datadog) {
+	rcService, isSet := rcServiceOptional.Get()
```
Comment (dustmop) [inline cmd/cluster-agent/subcommands/start/command.go:205]: We tend not to use "Optional" in variable names for optional types. Instead, I would recommend replacing `rcServiceOptional` with `rcService` and `rcService` with `rcserv` as a contraction is how this naming is usually handled (when an optional and a concrete type appear in the same function).

File: cmd/cluster-agent/subcommands/start/command.go [hunk 3]
```diff
@@ -198,17 +202,15 @@ func start(log log.Component, config config.Component, taggerComp tagger.Compone
...
+	if pkgconfig.IsRemoteConfigEnabled(pkgconfig.Datadog) && isSet {
```
Comment (dustmop) [inline cmd/cluster-agent/subcommands/start/command.go:206]: While you're here, this let's us remove a direct reference to the global config object, in favor the config component:

```suggestion
	if pkgconfig.IsRemoteConfigEnabled(config) && isSet {
```

**Review by dustmop — APPROVED** [review]
LGTM, just some small comments

**Pull request merged**